base_attachment_object_storage/models/ir_attachment.py

```diff
@@ -239,6 +267,65 @@ def force_storage(self):
             return super(IrAttachment, self).force_storage()
         self._force_storage_to_object_storage()
 
+    @api.model
+    def force_storage_to_db_for_special_fields(self, new_cr=False):
+        """Migrate special attachments from Object Storage back to database
+
+        The access to a file stored on the objects storage is slower
+        than a local disk or database access. For attachments like
+        image_small that are accessed in batch for kanban views, this
+        is too slow. We store this type of attachment in the database.
+
+        This method can be used when migrating a filestore where all the files,
+        including the special files (assets, image_small, ...) have been pushed
+        to the Object Storage and we want to write them back in the database.
+
+        It is not called anywhere, but can be called by RPC or scripts.
+        """
+        storage = self._storage()
+        if storage not in self._get_stores():
+            return
+
+        domain = AND((
+            normalize_domain(
+                [('store_fname', '=like', '{}://%'.format(storage))]
+            ),
+            normalize_domain(self._save_in_db_domain())
+        ))
+
+        with self.do_in_new_env(new_cr=new_cr) as new_env:
+            model_env = new_env['ir.attachment'].with_context(
+                prefetch_fields=False
+            )
+            attachment_ids = model_env.search(domain).ids
+            if not attachment_ids:
+                return
+            total = len(attachment_ids)
+            start_time = time.time()
+            _logger.info('Moving %d attachments from %s to'
+                         ' DB for fast access', total, storage)
+            current = 0
+            for attachment_id in attachment_ids:
+                current += 1
+                # if we browse attachments outside of the loop, the first
+                # access to 'datas' will compute all the 'datas' fields at
+                # once, which means reading hundreds or thousands of files at
+                # once, exhausting memory
+                attachment = model_env.browse(attachment_id)
+                # this write will read the datas from the Object Storage and
+                # write them back in the DB (the logic for location to write is
+                # in the 'datas' inverse computed field)
+                attachment.write({'datas': attachment.datas})
+                # as the file will potentially be dropped on the bucket,
+                # we should commit the changes here
```

Comment:
I don't get the link between the bucket and the change in DB

I don't understand the meaning of  "dropped on", is it:

A. A typo to say dropped from as in Deleted from the bucket?
B. To mean pushed into the bucket?

My guess is that it is B. but I don't get why. From what I understand we want to give shelter to the files into the database and not kick them out.
So it should be closer to A, but then do we really want to delete the files from S3 I would expect to not erase them right away. in this case I just spotted a typo.

Reply:
A. deleted from the bucket. This is how the inverse field works in ir_attachment.
Yes, there is a risk to lose the files if the commit fails. But all in all, the small/medium files can be rebuilt from the image so I'm not sure we want to make it too complex. (Deleting the files after would mean we have to keep a log of files to garbage collect)